### PR TITLE
Helm: Tidy up k8s versions checks

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
### Context 

Helm has built-in variables `.Capabilities.APIVersions` and `.Capabilities.KubeVersion.Version` that do what you think. Most of the time. We use these variables to decide what API versions to use for some resources or what fields to set on resources depending on what the user's Kubernetes version supports.

While working on https://github.com/grafana/mimir/pull/2870 I noticed that the built-in helm variables `.Capabilities.APIVersions` and `.Capabilities.KubeVersion.Version` depend on the helm command which is running.

If running `hem template` helm uses the kubectl client version and a trimmed down set of API resources.

```
[v1 admissionregistration.k8s.io/v1 admissionregistration.k8s.io/v1beta1 internal.apiserver.k8s.io/v1alpha1 apps/v1 apps/v1beta1 apps/v1beta2 authentication.k8s.io/v1 authentication.k8s.io/v1beta1 authorization.k8s.io/v1 authorization.k8s.io/v1beta1 autoscaling/v1 autoscaling/v2 autoscaling/v2beta1 autoscaling/v2beta2 batch/v1 batch/v1beta1 certificates.k8s.io/v1 certificates.k8s.io/v1beta1 coordination.k8s.io/v1beta1 coordination.k8s.io/v1 discovery.k8s.io/v1 discovery.k8s.io/v1beta1 events.k8s.io/v1 events.k8s.io/v1beta1 extensions/v1beta1 flowcontrol.apiserver.k8s.io/v1alpha1 flowcontrol.apiserver.k8s.io/v1beta1 flowcontrol.apiserver.k8s.io/v1beta2 networking.k8s.io/v1 networking.k8s.io/v1beta1 node.k8s.io/v1 node.k8s.io/v1alpha1 node.k8s.io/v1beta1 policy/v1 policy/v1beta1 rbac.authorization.k8s.io/v1 rbac.authorization.k8s.io/v1beta1 rbac.authorization.k8s.io/v1alpha1 scheduling.k8s.io/v1alpha1 scheduling.k8s.io/v1beta1 scheduling.k8s.io/v1 storage.k8s.io/v1beta1 storage.k8s.io/v1 storage.k8s.io/v1alpha1 apiextensions.k8s.io/v1beta1 apiextensions.k8s.io/v1]
```

When running `helm upgrade` helm uses the server version and the full list of API resources.
```
[apps/v1 v1/Eviction certificates.k8s.io/v1/CertificateSigningRequest v1/NodeProxyOptions v1/PersistentVolumeClaim apiregistration.k8s.io/v1/APIService metrics.k8s.io/v1beta1 traefik.containo.us/v1alpha1/IngressRoute rbac.authorization.k8s.io/v1/ClusterRole admissionregistration.k8s.io/v1/MutatingWebhookConfiguration flowcontrol.apiserver.k8s.io/v1beta2/PriorityLevelConfiguration autoscaling/v2beta2/HorizontalPodAutoscaler authorization.k8s.io/v1 v1/Node apps/v1/DaemonSet policy/v1/PodDisruptionBudget v1/TokenRequest apps/v1/ControllerRevision authorization.k8s.io/v1/LocalSubjectAccessReview traefik.containo.us/v1alpha1/MiddlewareTCP flowcontrol.apiserver.k8s.io/v1beta1/PriorityLevelConfiguration helm.cattle.io/v1/HelmChart traefik.containo.us/v1alpha1/TLSOption v1/ServiceProxyOptions scheduling.k8s.io/v1/PriorityClass v1/LimitRange v1/PodTemplate discovery.k8s.io/v1 authorization.k8s.io/v1/SelfSubjectRulesReview v1/Secret apps/v1/ReplicaSet batch/v1/CronJob storage.k8s.io/v1/VolumeAttachment authentication.k8s.io/v1 batch/v1/Job rbac.authorization.k8s.io/v1/ClusterRoleBinding rbac.authorization.k8s.io/v1/Role apiextensions.k8s.io/v1/CustomResourceDefinition v1 certificates.k8s.io/v1 v1/Endpoints networking.k8s.io/v1/IngressClass storage.k8s.io/v1/CSIDriver storage.k8s.io/v1/StorageClass metrics.k8s.io/v1beta1/NodeMetrics metrics.k8s.io/v1beta1/PodMetrics events.k8s.io/v1 networking.k8s.io/v1 flowcontrol.apiserver.k8s.io/v1beta1 rbac.authorization.k8s.io/v1 admissionregistration.k8s.io/v1 helm.cattle.io/v1 policy/v1 traefik.containo.us/v1alpha1/ServersTransport autoscaling/v2 storage.k8s.io/v1beta1 authorization.k8s.io/v1/SubjectAccessReview flowcontrol.apiserver.k8s.io/v1beta2 networking.k8s.io/v1/NetworkPolicy autoscaling/v2/HorizontalPodAutoscaler k3s.cattle.io/v1 v1/PersistentVolume v1/PodAttachOptions coordination.k8s.io/v1/Lease traefik.containo.us/v1alpha1/TLSStore v1/ComponentStatus rbac.authorization.k8s.io/v1/RoleBinding k3s.cattle.io/v1/Addon authentication.k8s.io/v1/TokenReview helm.cattle.io/v1/HelmChartConfig traefik.containo.us/v1alpha1/IngressRouteUDP v1/ResourceQuota discovery.k8s.io/v1/EndpointSlice v1/ConfigMap v1/Event v1/Scale node.k8s.io/v1 storage.k8s.io/v1/CSINode v1/PodProxyOptions apps/v1/StatefulSet autoscaling/v1/HorizontalPodAutoscaler apps/v1/Scale traefik.containo.us/v1alpha1/Middleware apiextensions.k8s.io/v1 admissionregistration.k8s.io/v1/ValidatingWebhookConfiguration autoscaling/v2beta2 batch/v1 traefik.containo.us/v1alpha1/IngressRouteTCP v1/PodExecOptions storage.k8s.io/v1beta1/CSIStorageCapacity traefik.containo.us/v1alpha1/TraefikService v1/ReplicationController flowcontrol.apiserver.k8s.io/v1beta2/FlowSchema scheduling.k8s.io/v1 v1/Namespace v1/Pod v1/Binding v1/Service events.k8s.io/v1/Event node.k8s.io/v1/RuntimeClass coordination.k8s.io/v1 v1/ServiceAccount apps/v1/Deployment networking.k8s.io/v1/Ingress flowcontrol.apiserver.k8s.io/v1beta1/FlowSchema autoscaling/v1 storage.k8s.io/v1 authorization.k8s.io/v1/SelfSubjectAccessReview storage.k8s.io/v1/CSIStorageCapacity apiregistration.k8s.io/v1 traefik.containo.us/v1alpha1 v1/PodPortForwardOptions]
```

`.Capabilities.APIVersions` can be configured by passing e.g. `--api-version=autoscaling/v2/HorizontalPodAutoscaler` to `helm template`. But I consider this bad user experience - asking users to specify which resources their cluster supports so the chart can render them. And keeping this list of resources up to date as the chart evolves. I don't think users should be concerned with version of all resources. `helm template` is also used by [helm-in-jsonnet](https://tanka.dev/helm) and probably in users' GitOps workflows.

edit: forgot to link the example code i used to show the behaviour - fdc23efae057ef83eb91f38206436e2f774ee14f

### Proposal

Only rely on the kubernetes version instead of the API resources. We already have a mechanism for overriding this from the values.yaml. This PR also documents it.

As part of this cleanup I discovered that the ingress resources use some checks for versions older than 1.19, which we don't support anyways, so I simplified the templates.

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: operations/helm/charts/mimir-distributed/templates/gateway/gateway-ingress.yaml [hunk 1]
```diff
@@ -1,9 +1,6 @@
 {{- if .Values.enterprise.enabled -}}
 {{- if .Values.gateway.ingress.enabled -}}
-{{- $ingressApiIsStable := eq (include "mimir.ingress.isStable" .) "true" -}}
-{{- $ingressSupportsIngressClassName := eq (include "mimir.ingress.supportsIngressClassName" .) "true" -}}
-{{- $ingressSupportsPathType := eq (include "mimir.ingress.supportsPathType" .) "true" -}}
-apiVersion: {{ include "mimir.ingress.apiVersion" . }}
+apiVersion: networking.k8s.io/v1
```
Comment (krajorama) [inline operations/helm/charts/mimir-distributed/templates/gateway/gateway-ingress.yaml:3]: Let's add a CHANGELOG entry `[CHANGE]` to highlight that ingress will not render on pre 1.19 kubernetes.

File: operations/helm/charts/mimir-distributed/templates/nginx/ingress.yaml [hunk 2]
```diff
@@ -1,10 +1,7 @@
 {{- if not .Values.enterprise.enabled -}}
 {{- if .Values.nginx.enabled -}}
 {{- if .Values.nginx.ingress.enabled -}}
-{{- $ingressApiIsStable := eq (include "mimir.ingress.isStable" .) "true" -}}
-{{- $ingressSupportsIngressClassName := eq (include "mimir.ingress.supportsIngressClassName" .) "true" -}}
-{{- $ingressSupportsPathType := eq (include "mimir.ingress.supportsPathType" .) "true" -}}
-apiVersion: {{ include "mimir.ingress.apiVersion" . }}
+apiVersion: networking.k8s.io/v1
```
Comment (krajorama) [inline operations/helm/charts/mimir-distributed/templates/nginx/ingress.yaml:4]: same as above

**Review by krajorama — APPROVED** [review]
Nit, people may have forced to use the chart on <1.20 before, now that's not an option, let's add to the changelog

File: operations/helm/charts/mimir-distributed/templates/gateway/gateway-ingress.yaml (hunk 1 above)
Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/templates/gateway/gateway-ingress.yaml:3]: addresseed in https://github.com/grafana/mimir/pull/2872/commits/f793324039e2460480df7205916e6236b0da4df8

**Pull request merged**